### PR TITLE
Add support for workers

### PR DESCRIPTION
**Pull request by clelland**
This adds support for configuring reporting in worker scripts, and for
generating reports from a worker context. Initialization of reporting endpoints
is decoupled from fetch so that it can be referenced in HTML instead, when
documents and workers are created.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/reporting/pull/206.html" title="Last updated on Oct 19, 2020, 1:41 PM UTC (26189fc)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/reporting/206/f890dc1...26189fc.html" title="Last updated on Oct 19, 2020, 1:41 PM UTC (26189fc)">Diff</a>

### CODE REVIEW

**Comment by clelland** [comment]
@bakulf, FYI
Also, this change should be in conjunction with https://github.com/whatwg/html/pull/5470, which handles timing of initializing the reporting endpoints in both documents and workers.

File: index.src.html [hunk 1]
```diff
@@ -495,24 +511,28 @@ <h4 id="generate-report" export algorithm>Generate report of |type| with
...
 
   When the user agent is to <dfn>generate a report</dfn> for a <a>Document</a>
-  (|document|), given a string (|type|), another string (|destination|), and
-  a serializable object (|data|), it must run the following steps:
+  or {{WorkerGlobalScope}} object (|context|), given a string (|type|), another
+  string (|destination|), and a serializable object (|data|), it must run the
+  following steps:
+
+  1.  If |context| is a <a>Document</a>, then let |settings| be |context|'s
+      [=environment settings object=].
```
Comment (domenic) [inline index.src.html:519]: relevant settings object should work for both. This link just goes to the general definition of environment settings object, which isn't right.

Comment (clelland) [reply index.src.html]: Thanks, updated

File: index.src.html [hunk 2]
```diff
@@ -330,11 +333,11 @@ <h3 id="queue-report" algorithm>
 
   5.  If |settings| is given, then
 
-      1.  Let |environment| be |settings|'s <a>realm execution context</a>'s
-          <a spec=ecmascript>realm</a>'s <a spec=ecmascript lt=realm>ECMAScript
-          global environment</a>.
+      1.  Let |scope| be |settings|'s <a>realm execution context</a>'s
+          <a spec=ecmascript>realm</a>'s <a spec=ecmascript>global object</a>.
```
Comment (domenic) [inline index.src.html:337]: This should just be `|setting|'s [=environment settings object/global object=]`

File: index.src.html [hunk 3]
```diff
@@ -390,27 +393,34 @@ <h3 id="serialize-reports" algorithm>Serialize Reports</h3>
...
-  via the <a>`Reporting-Endpoints`</a> HTTP response header field. This
-  mechanism is defined in [[#header]], and its processing in
-  [[#process-header]].
+  Each object implementing {{WindowOrWorkerGlobalScope}} has an <dfn
+  for="WindowOrWorkerGlobalScope" export>reports</dfn> list, which is
+  a list of <a for="/">reports</a>.
 
-  Each <a>document</a> has an <dfn for="document" export attribute>reports</dfn>
-  list, which is a list of <a>reports</a>.
+  To <dfn export>initialize a document or worker's endpoint list</dfn>, given a
```
Comment (domenic) [inline index.src.html:412]: Should this be "initialize a global's endpoint list" perhaps?

File: index.src.html [hunk 4]
```diff
@@ -522,14 +527,14 @@ <h3 id="report-delivery">Report Delivery</h3>
...
+  or {{WorkerGlobalScope}} object (|context|) by executing the following steps:
 
   1.  Let |endpoint map| be an empty map of <a>endpoint</a> objects to lists of
       <a>report</a> objects.
 
   2.  For each |report| in |reports|:
 
-      1.  If there exists an <a>endpoint</a> (|endpoint|) in |document|'s
+      1.  If there exists an <a>endpoint</a> (|endpoint|) in |context|'s
           {{document/endpoints}} list whose {{endpoint/name}} is |report|'s
```
Comment (domenic) [inline index.src.html:538]: This seems like it needs updating?

File: index.src.html [hunk 5]
```diff
@@ -561,7 +566,7 @@ <h4 id="send-reports" algorithm>Send reports</h4>
 
           3.  If |result| is "`Remove Endpoint`":
 
-              1.  Remove |endpoint| from |document|'s {{document/endpoints}}
+              1.  Remove |endpoint| from |context|'s {{document/endpoints}}
                   list.
```
Comment (domenic) [inline index.src.html:570]: Similarly, any reference to `{{document/endpoints}}` seems like it needs updating.

File: index.src.html [hunk 6]
```diff
@@ -728,29 +733,29 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
   The <dfn method for=ReportingObserver><code>observe()</code></dfn>
   method, when invoked, must run these steps:
 
-  1. Let |environment| be the <a spec=ecmascript lt=realm>ECMAScript global
-     environment</a> associated with the <a>context object</a>.
+  1. Let |global| be the <a spec=ecmascript>global object</a> of the <a
+     spec=ecmascript>realm</a> associated with the <a>context object</a>.
```
Comment (domenic) [inline index.src.html:737]: This can be `Let |global| be the [=relevant global object=] of [=this=]`. (Although perhaps you should keep using "context object" for consistency for now, instead of "this".)

File: index.src.html [hunk 7]
```diff
@@ -728,29 +733,29 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
   The <dfn method for=ReportingObserver><code>disconnect()</code></dfn> method,
   when invoked, must run these steps:
 
   1. If the <a>context object</a> is not <a>registered</a>, return.
 
-  2. Remove the <a>context object</a> from the <a>registered reporting observer
-     list</a> of its associated <a spec=ecmascript lt=realm>ECMAScript global
-     environment</a>.
+  2. Let |global| be the <a spec=ecmascript>global object</a> of the <a
+     spec=ecmascript>realm</a> associated with the <a>context object</a>.
```
Comment (domenic) [inline index.src.html:755]: Same as above.

File: index.src.html [hunk 8]
```diff
@@ -803,10 +807,11 @@ <h3 id="add-report" algorithm>
 
   3. Append |r| to |observer|'s <a>report queue</a>.
 
-  4. If the size of |observer|'s <a>report queue</a> is 1, <a>Queue a task</a>
-     to [[#invoke-observers]] with a copy of the <a>registered reporting
-     observer list</a> of the <a spec=ecmascript lt=realm>ECMAScript global
-     environment</a> associated with |observer|.
+  4. If the size of |observer|'s <a>report queue</a> is 1:
+     1.  Let |global| be the <a spec=ecmascript>global object</a> of the <a
+         spec=ecmascript>realm</a> associated with |observer|.
```
Comment (domenic) [inline index.src.html:812]: Same as above.

File: index.src.html [hunk 9]
```diff
@@ -803,10 +807,11 @@ <h3 id="add-report" algorithm>
...
+     2.  <a>Queue a task</a> to [[#invoke-observers]] with a copy of |global|'s
+         <a>registered reporting observer list</a>.
```
Comment (domenic) [inline index.src.html:814]: Drive-by preexisting issue: These days we don't like bare "Queue a task". This should be "Queue a global task on the XXX task source given |global| to ...".

File: index.src.html (hunk 3 above)
Comment (domenic) [reply index.src.html:412]: I'm assuming the idea is to patch HTML to run this algorithm on global object creation?

Comment (clelland) [reply index.src.html:412]: Yes, that's in progress in https://github.com/whatwg/html/pull/5470

File: index.src.html (hunk 2 above)
Comment (clelland) [reply index.src.html:337]: I would have thought so, but [HTML§8.1.3.2](https://html.spec.whatwg.org/multipage/webappapis.html#environment-settings-objects) doesn't define it, and [HTML§8.1.3.3](https://html.spec.whatwg.org/multipage/webappapis.html#realms-settings-objects-global-objects) explains this roundabout way of getting to it.

File: index.src.html (hunk 4 above)
Comment (clelland) [reply index.src.html]: Thanks, yes it does.

File: index.src.html (hunk 5 above)
Comment (clelland) [reply index.src.html]: I searched for "endpoints<", but missed these :( Thanks for catching.

File: index.src.html (hunk 6 above)
Comment (clelland) [reply index.src.html]: Done. (Will switch to [=this=] soon)

File: index.src.html (hunk 9 above)
Comment (clelland) [reply index.src.html:814]: Thanks -- filed #220.

File: index.src.html (hunk 8 above)
Comment (clelland) [reply index.src.html]: done.

File: index.src.html (hunk 7 above)
Comment (clelland) [reply index.src.html]: done.

File: index.src.html (hunk 2 above)
Comment (domenic) [reply index.src.html:337]: I'm not sure I understand the response... HTML's https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-global is meant to be something other specs can reference when they want to go from ESOs to global objects.

Comment (clelland) [reply index.src.html:337]: Ahh, now I'm reading it right -- I didn't realize that the text at the end was a definition; I thought that it was explaining how to get to the global object by going through the Realm.

I was looking for something like "An _environment settings object_ has a **global object**, which is..."

Comment (clelland) [reply index.src.html]: Done!

File: index.src.html (hunk 3 above)
Comment (domenic) [reply index.src.html:408]: Ping on

> Should this be "initialize a global's endpoint list" perhaps?

**Review by domenic — APPROVED** [review]
LGTM modulo algorithm naming

File: index.src.html (hunk 3 above)
Comment (clelland) [reply index.src.html:408]: > Should this be "initialize a global's endpoint list" perhaps?

I missed that one; that's a better name 👍
